File: legacy/backend/scheduler.py
from datetime import date, timedelta
from typing import List, Dict, Set, Optional
from collections import defaultdict
import random

from .models import Doctor, Shift, ShiftType, AppState # Removed DoctorAffinity

def get_italian_holidays_and_prefestivi(year: int) -> (List[date], List[date]):
    """
    Returns a list of Italian holidays and prefestivi for the given year.
    """
    holidays = [
        date(year, 1, 1), date(year, 1, 6), date(year, 4, 25), date(year, 5, 1),
        date(year, 6, 2), date(year, 8, 15), date(year, 11, 1), date(year, 12, 8),
        date(year, 12, 25), date(year, 12, 26)
    ]
    # Easter calculation
    a = year % 19; b = year % 4; c = year % 7; k = year // 100; p = (13 + 8 * k) // 25
    q = k // 4; M = (15 - p + k - q) % 30; N = (4 + k - q) % 7
    d = (19 * a + M) % 30; e = (2 * b + 4 * c + 6 * d + N) % 7
    easter_day = 22 + d + e
    easter_month = 3
    if easter_day > 31:
        easter_day -= 31
        easter_month = 4
    
    easter = date(year, easter_month, easter_day)
    holidays.append(easter)
    holidays.append(easter + timedelta(days=1)) # Pasquetta
    holidays = sorted(list(set(holidays)))

    prefestivi = []
    for holiday in holidays:
        prefestivo = holiday - timedelta(days=1)
        # A prefestivo cannot be a holiday itself or a Sunday (weekday() == 6)
        if prefestivo not in holidays and prefestivo.weekday() != 6:
            prefestivi.append(prefestivo)
            
    return holidays, sorted(list(set(prefestivi)))

def generate_shifts_for_month(year: int, month: int, holidays: List[date], prefestivi: List[date]) -> List[Shift]:
    """
    Generates all shifts for a month based on the corrected rules:
    - Feriale: Only one SERALE shift.
    - Sabato/Domenica/Festivi/Prefestivi: GIORNO and NOTTE shifts.
    All shifts require 2 doctors.
    """
    shifts: List[Shift] = []
    current_date = date(year, month, 1)
    
    while current_date.month == month:
        day_of_week = current_date.weekday() # Monday is 0, Sunday is 6
        is_holiday = current_date in holidays
        is_prefestivo = current_date in prefestivi

        # Weekend, Holiday, or day before a holiday
        if is_holiday or is_prefestivo or day_of_week in [5, 6]:
            if day_of_week == 5 or is_prefestivo: # Sabato or Prefestivo
                shifts.append(Shift(shift_id=f"{current_date.isoformat()}_giorno", shift_date=current_date, shift_type=ShiftType.SABATO_GIORNO, max_doctors=2))
                shifts.append(Shift(shift_id=f"{current_date.isoformat()}_notte", shift_date=current_date, shift_type=ShiftType.SABATO_NOTTE, max_doctors=2))
            else: # Domenica or Festivo
                shifts.append(Shift(shift_id=f"{current_date.isoformat()}_giorno", shift_date=current_date, shift_type=ShiftType.DOMENICA_GIORNO, max_doctors=2))
                shifts.append(Shift(shift_id=f"{current_date.isoformat()}_notte", shift_date=current_date, shift_type=ShiftType.DOMENICA_NOTTE, max_doctors=2))
        else: # Feriale
            shifts.append(Shift(shift_id=f"{current_date.isoformat()}_feriale_serale", shift_date=current_date, shift_type=ShiftType.FERIALE_SERALE, max_doctors=2))

        current_date += timedelta(days=1)
    return shifts

def is_evening_shift(shift_type: ShiftType) -> bool:
    """Checks if a shift is in the evening/night."""
    return shift_type.name.endswith('_NOTTE') or shift_type == ShiftType.FERIALE_SERALE

def is_contiguous(shift1: Shift, shift2: Shift) -> bool:
    """
    Determines if two shifts are contiguous.
    - Same day: day shift and evening shift are contiguous.
    - Consecutive days: evening shift and next day's day shift are contiguous.
    """
    # Ensure shift1 is before shift2 for easier comparison
    if shift1.shift_date > shift2.shift_date:
        shift1, shift2 = shift2, shift1
        
    shift1_is_evening = is_evening_shift(shift1.shift_type)
    shift2_is_evening = is_evening_shift(shift2.shift_type)

    # Case 1: Same day. Contiguous if one is evening and the other is not.
    if shift1.shift_date == shift2.shift_date:
        return shift1_is_evening != shift2_is_evening

    # Case 2: Consecutive days. Contiguous if the first is evening and the second is day.
    if (shift1.shift_date + timedelta(days=1)) == shift2.shift_date:
        return shift1_is_evening and not shift2_is_evening
        
    return False

def count_consecutive_nights(doctor_id: str, candidate_shift: Shift, assignments: Dict[str, List[Shift]]) -> int:
    """
    Count how many consecutive night shifts this assignment would create.
    Returns the length of the consecutive-night run including the candidate.
    """
    if not is_evening_shift(candidate_shift.shift_type):
        return 0

    night_dates: Set[date] = set()
    for s in assignments[doctor_id]:
        if is_evening_shift(s.shift_type):
            night_dates.add(s.shift_date)

    candidate_date = candidate_shift.shift_date
    night_dates.add(candidate_date)

    # Walk backwards and forwards to measure the full consecutive run
    count = 1
    d = candidate_date - timedelta(days=1)
    while d in night_dates:
        count += 1
        d -= timedelta(days=1)
    d = candidate_date + timedelta(days=1)
    while d in night_dates:
        count += 1
        d += timedelta(days=1)

    return count

def assign_shifts(doctors: List[Doctor], shifts: List[Shift]) -> List[Shift]:
    doctor_assignments: Dict[str, List[Shift]] = defaultdict(list)
    doctors_by_id: Dict[str, Doctor] = {d.doctor_id: d for d in doctors}

    # --- Cost Function ---
    def get_assignment_cost(doctor: Doctor, shift: Shift, assignments: Dict[str, List[Shift]]) -> float:
        # --- Hard Constraints (Infinite Cost) ---
        if len(assignments[doctor.doctor_id]) >= doctor.max_shifts:
            return float('inf')
        if shift.shift_date in doctor.indisponibilita:
            return float('inf')
        if any(is_contiguous(s, shift) for s in assignments[doctor.doctor_id]):
            return float('inf')

        # Hard block: never allow 3+ consecutive nights
        consec = count_consecutive_nights(doctor.doctor_id, shift, assignments)
        if consec >= 3:
            return float('inf')

        cost = 100.0  # Base cost for any assignment

        # --- Soft Constraints (Cost Modifiers) ---

        # 1. Consecutive nights penalty: 2 consecutive nights is heavily penalised
        if consec == 2:
            cost += 60

        # 2. Shift Type Distribution
        doc_shifts = assignments[doctor.doctor_id]
        num_feriale = sum(1 for s in doc_shifts if s.shift_type == ShiftType.FERIALE_SERALE)
        num_sabato_g = sum(1 for s in doc_shifts if s.shift_type == ShiftType.SABATO_GIORNO)
        num_sabato_n = sum(1 for s in doc_shifts if s.shift_type == ShiftType.SABATO_NOTTE)
        num_domenica_g = sum(1 for s in doc_shifts if s.shift_type == ShiftType.DOMENICA_GIORNO)
        num_domenica_n = sum(1 for s in doc_shifts if s.shift_type == ShiftType.DOMENICA_NOTTE)

        # Add cost for exceeding preferred counts, bonus for meeting them
        if shift.shift_type == ShiftType.FERIALE_SERALE:
            if num_feriale < doctor.preferred_feriali: cost -= 20 # Bonus
            else: cost += (num_feriale - doctor.preferred_feriali + 1) * 10 # Penalty
        elif shift.shift_type == ShiftType.SABATO_GIORNO:
            if num_sabato_g < doctor.preferred_sabato_giorno: cost -= 30
            else: cost += (num_sabato_g - doctor.preferred_sabato_giorno + 1) * 20
        elif shift.shift_type == ShiftType.SABATO_NOTTE:
            if num_sabato_n < doctor.preferred_sabato_notte: cost -= 30
            else: cost += (num_sabato_n - doctor.preferred_sabato_notte + 1) * 20
        elif shift.shift_type == ShiftType.DOMENICA_GIORNO:
            if num_domenica_g < doctor.preferred_domenica_giorno: cost -= 30
            else: cost += (num_domenica_g - doctor.preferred_domenica_giorno + 1) * 20
        elif shift.shift_type == ShiftType.DOMENICA_NOTTE:
            if num_domenica_n < doctor.preferred_domenica_notte: cost -= 30
            else: cost += (num_domenica_n - doctor.preferred_domenica_notte + 1) * 20

        # 3. Affinity bonus (bidirectional, strong weight)
        if len(shift.assigned_doctor_ids) > 0:
            first_doctor_id = shift.assigned_doctor_ids[0]
            first_doctor_obj = doctors_by_id.get(first_doctor_id)

            # First doctor prefers this candidate
            if first_doctor_obj and doctor.doctor_id in first_doctor_obj.preferred_colleagues:
                rank = first_doctor_obj.preferred_colleagues.index(doctor.doctor_id)
                max_rank = len(first_doctor_obj.preferred_colleagues)
                cost -= 50 * (max_rank - rank) / max_rank

            # This candidate prefers the first doctor (bidirectional)
            if first_doctor_id in doctor.preferred_colleagues:
                rank = doctor.preferred_colleagues.index(first_doctor_id)
                max_rank = len(doctor.preferred_colleagues)
                cost -= 50 * (max_rank - rank) / max_rank

        # 4. Add a small random factor to break ties
        cost += random.uniform(0, 0.9)
        return cost

    # --- Main Assignment Loop ---
    # Loop until all shifts are filled or no more assignments are possible
    while True:
        unfilled_slots = []
        for s in shifts:
            for _ in range(s.max_doctors - len(s.assigned_doctor_ids)):
                unfilled_slots.append(s)
        
        if not unfilled_slots:
            break # All shifts are full

        best_assignment = None
        min_cost = float('inf')

        # Find the single best (lowest cost) assignment across all possibilities
        for shift_to_fill in unfilled_slots:
            eligible_doctors = [d for d in doctors if d.doctor_id not in shift_to_fill.assigned_doctor_ids]
            for doctor in eligible_doctors:
                cost = get_assignment_cost(doctor, shift_to_fill, doctor_assignments)
                if cost < min_cost:
                    min_cost = cost
                    best_assignment = (doctor, shift_to_fill)

        # If no valid assignment was found in the entire pass, break
        if min_cost == float('inf') or best_assignment is None:
            break
            
        # Make the best assignment found
        best_doctor, best_shift = best_assignment
        best_shift.assigned_doctor_ids.append(best_doctor.doctor_id)
        doctor_assignments[best_doctor.doctor_id].append(best_shift)

    # Final check for warnings
    for shift in shifts:
        if len(shift.assigned_doctor_ids) < shift.max_doctors:
            print(f"Warning: Could not find any eligible doctor for a remaining slot in shift {shift.shift_id}. The slot remains unfilled.")
            
    return shifts

def generate_schedule(doctors: List[Doctor], year: int, month: int) -> AppState: # Removed affinities
    holidays, prefestivi = get_italian_holidays_and_prefestivi(year)
    all_shifts = generate_shifts_for_month(year, month, holidays, prefestivi)
    
    assigned_shifts = assign_shifts(doctors, all_shifts) # Removed affinities
    
    return AppState(
        doctors=doctors,
        shifts=assigned_shifts,
        holidays=holidays
        # Removed affinities
    )

File: app/src/lib/scheduler.ts
// Scheduler — ported from Python backend with improved constraints

import { ShiftType, type Doctor, type Shift } from './models';

/* ---------- Italian Holidays & Prefestivi ---------- */

function computeEaster(year: number): Date {
  const a = year % 19;
  const b = year % 4;
  const c = year % 7;
  const k = Math.floor(year / 100);
  const p = Math.floor((13 + 8 * k) / 25);
  const q = Math.floor(k / 4);
  const M = (15 - p + k - q) % 30;
  const N = (4 + k - q) % 7;
  const d = (19 * a + M) % 30;
  const e = (2 * b + 4 * c + 6 * d + N) % 7;
  let easterDay = 22 + d + e;
  let easterMonth = 2; // 0-indexed March
  if (easterDay > 31) {
    easterDay -= 31;
    easterMonth = 3; // April
  }
  return new Date(year, easterMonth, easterDay);
}

function toISO(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

function addDays(d: Date, n: number): Date {
  const r = new Date(d);
  r.setDate(r.getDate() + n);
  return r;
}

export function getItalianHolidaysAndPrefestivi(year: number): {
  holidays: string[];
  prefestivi: string[];
} {
  const holidayDates: Date[] = [
    new Date(year, 0, 1),
    new Date(year, 0, 6),
    new Date(year, 3, 25),
    new Date(year, 4, 1),
    new Date(year, 5, 2),
    new Date(year, 7, 15),
    new Date(year, 10, 1),
    new Date(year, 11, 8),
    new Date(year, 11, 25),
    new Date(year, 11, 26),
  ];

  const easter = computeEaster(year);
  holidayDates.push(easter);
  holidayDates.push(addDays(easter, 1)); // Pasquetta

  const holidaySet = new Set(holidayDates.map(toISO));
  const holidays = Array.from(holidaySet).sort();

  const prefestivi: string[] = [];
  for (const h of holidayDates) {
    const pf = addDays(h, -1);
    const pfISO = toISO(pf);
    if (!holidaySet.has(pfISO) && pf.getDay() !== 0) {
      // Not a holiday itself and not a Sunday
      prefestivi.push(pfISO);
    }
  }
  return { holidays, prefestivi: [...new Set(prefestivi)].sort() };
}

/* ---------- Shift Generation ---------- */

export function generateShiftsForMonth(
  year: number,
  month: number, // 1-based
  holidays: string[],
  prefestivi: string[]
): Shift[] {
  const shifts: Shift[] = [];
  const holidaySet = new Set(holidays);
  const prefestivoSet = new Set(prefestivi);

  const daysInMonth = new Date(year, month, 0).getDate();

  for (let day = 1; day <= daysInMonth; day++) {
    const d = new Date(year, month - 1, day);
    const iso = toISO(d);
    const dow = d.getDay(); // 0=Sun, 6=Sat
    const isHoliday = holidaySet.has(iso);
    const isPrefestivo = prefestivoSet.has(iso);

    if (isHoliday || isPrefestivo || dow === 0 || dow === 6) {
      if (dow === 6 || isPrefestivo) {
        shifts.push({
          shift_id: `${iso}_giorno`,
          shift_date: iso,
          shift_type: ShiftType.SABATO_GIORNO,
          assigned_doctor_ids: [],
          max_doctors: 2,
        });
        shifts.push({
          shift_id: `${iso}_notte`,
          shift_date: iso,
          shift_type: ShiftType.SABATO_NOTTE,
          assigned_doctor_ids: [],
          max_doctors: 2,
        });
      } else {
        shifts.push({
          shift_id: `${iso}_giorno`,
          shift_date: iso,
          shift_type: ShiftType.DOMENICA_GIORNO,
          assigned_doctor_ids: [],
          max_doctors: 2,
        });
        shifts.push({
          shift_id: `${iso}_notte`,
          shift_date: iso,
          shift_type: ShiftType.DOMENICA_NOTTE,
          assigned_doctor_ids: [],
          max_doctors: 2,
        });
      }
    } else {
      shifts.push({
        shift_id: `${iso}_feriale_serale`,
        shift_date: iso,
        shift_type: ShiftType.FERIALE_SERALE,
        assigned_doctor_ids: [],
        max_doctors: 2,
      });
    }
  }
  return shifts;
}

/* ---------- Helpers ---------- */

export function isEveningShift(st: ShiftType): boolean {
  return st.endsWith('_notte') || st === ShiftType.FERIALE_SERALE;
}

export function isContiguous(s1: Shift, s2: Shift): boolean {
  let a = s1,
    b = s2;
  if (a.shift_date > b.shift_date) {
    [a, b] = [b, a];
  }

  const aEvening = isEveningShift(a.shift_type);
  const bEvening = isEveningShift(b.shift_type);

  // Same day: contiguous if one is day and one is evening
  if (a.shift_date === b.shift_date) {
    return aEvening !== bEvening;
  }

  // Consecutive days: evening → next day's day shift
  const aDate = new Date(a.shift_date);
  const bDate = new Date(b.shift_date);
  const diffMs = bDate.getTime() - aDate.getTime();
  if (Math.round(diffMs / 86400000) === 1) {
    return aEvening && !bEvening;
  }

  return false;
}

function countConsecutiveNights(
  doctorId: string,
  candidateShift: Shift,
  assignments: Map<string, Shift[]>
): number {
  if (!isEveningShift(candidateShift.shift_type)) return 0;

  const nightDates = new Set<string>();
  for (const s of assignments.get(doctorId) ?? []) {
    if (isEveningShift(s.shift_type)) nightDates.add(s.shift_date);
  }
  nightDates.add(candidateShift.shift_date);

  const cand = new Date(candidateShift.shift_date);
  let count = 1;
  let d = addDays(cand, -1);
  while (nightDates.has(toISO(d))) {
    count++;
    d = addDays(d, -1);
  }
  d = addDays(cand, 1);
  while (nightDates.has(toISO(d))) {
    count++;
    d = addDays(d, 1);
  }
  return count;
}

function countShiftsInWeek(
  doctorId: string,
  candidateShift: Shift,
  assignments: Map<string, Shift[]>
): number {
  const candDate = new Date(candidateShift.shift_date);
  const windowStart = addDays(candDate, -3);
  const windowEnd = addDays(candDate, 3);
  const startISO = toISO(windowStart);
  const endISO = toISO(windowEnd);

  let count = 1; // counting the candidate itself
  for (const s of assignments.get(doctorId) ?? []) {
    if (s.shift_date >= startISO && s.shift_date <= endISO) {
      count++;
    }
  }
  return count;
}

/* ---------- Assignment Algorithm ---------- */

export function assignShifts(doctors: Doctor[], shifts: Shift[]): Shift[] {
  const assignments = new Map<string, Shift[]>();
  for (const d of doctors) assignments.set(d.doctor_id, []);

  const doctorsById = new Map(doctors.map((d) => [d.doctor_id, d]));

  function getCost(doctor: Doctor, shift: Shift): number {
    const docShifts = assignments.get(doctor.doctor_id) ?? [];

    // --- Hard constraints ---
    if (docShifts.length >= doctor.max_shifts) return Infinity;
    if (doctor.indisponibilita.includes(shift.shift_date)) return Infinity;
    if (docShifts.some((s) => isContiguous(s, shift))) return Infinity;

    const consec = countConsecutiveNights(doctor.doctor_id, shift, assignments);
    if (consec >= 3) return Infinity;

    let cost = 100;

    // --- Soft: 2 consecutive nights penalty ---
    if (consec === 2) cost += 60;

    // --- Soft: weekly spread — penalise clustering ---
    const weekCount = countShiftsInWeek(doctor.doctor_id, shift, assignments);
    if (weekCount > 3) cost += (weekCount - 3) * 25;

    // --- Soft: shift type distribution ---
    const counts = {
      feriale: 0,
      sabato_g: 0,
      sabato_n: 0,
      domenica_g: 0,
      domenica_n: 0,
    };
    for (const s of docShifts) {
      if (s.shift_type === ShiftType.FERIALE_SERALE) counts.feriale++;
      else if (s.shift_type === ShiftType.SABATO_GIORNO) counts.sabato_g++;
      else if (s.shift_type === ShiftType.SABATO_NOTTE) counts.sabato_n++;
      else if (s.shift_type === ShiftType.DOMENICA_GIORNO) counts.domenica_g++;
      else if (s.shift_type === ShiftType.DOMENICA_NOTTE) counts.domenica_n++;
    }

    if (shift.shift_type === ShiftType.FERIALE_SERALE) {
      cost += counts.feriale < doctor.preferred_feriali ? -20 : (counts.feriale - doctor.preferred_feriali + 1) * 10;
    } else if (shift.shift_type === ShiftType.SABATO_GIORNO) {
      cost += counts.sabato_g < doctor.preferred_sabato_giorno ? -30 : (counts.sabato_g - doctor.preferred_sabato_giorno + 1) * 20;
    } else if (shift.shift_type === ShiftType.SABATO_NOTTE) {
      cost += counts.sabato_n < doctor.preferred_sabato_notte ? -30 : (counts.sabato_n - doctor.preferred_sabato_notte + 1) * 20;
    } else if (shift.shift_type === ShiftType.DOMENICA_GIORNO) {
      cost += counts.domenica_g < doctor.preferred_domenica_giorno ? -30 : (counts.domenica_g - doctor.preferred_domenica_giorno + 1) * 20;
    } else if (shift.shift_type === ShiftType.DOMENICA_NOTTE) {
      cost += counts.domenica_n < doctor.preferred_domenica_notte ? -30 : (counts.domenica_n - doctor.preferred_domenica_notte + 1) * 20;
    }

    // --- Soft: bidirectional affinity bonus (strong weight) ---
    if (shift.assigned_doctor_ids.length > 0) {
      const firstId = shift.assigned_doctor_ids[0];
      const firstDoc = doctorsById.get(firstId);

      if (firstDoc && firstDoc.preferred_colleagues.includes(doctor.doctor_id)) {
        const rank = firstDoc.preferred_colleagues.indexOf(doctor.doctor_id);
        const max = firstDoc.preferred_colleagues.length;
        cost -= 50 * ((max - rank) / max);
      }
      if (doctor.preferred_colleagues.includes(firstId)) {
        const rank = doctor.preferred_colleagues.indexOf(firstId);
        const max = doctor.preferred_colleagues.length;
        cost -= 50 * ((max - rank) / max);
      }
    }

    // Tie-breaker
    cost += Math.random() * 0.9;
    return cost;
  }

  // Greedy loop
  while (true) {
    const unfilled: Shift[] = [];
    for (const s of shifts) {
      for (let i = s.assigned_doctor_ids.length; i < s.max_doctors; i++) {
        unfilled.push(s);
      }
    }
    if (unfilled.length === 0) break;

    let bestCost = Infinity;
    let bestPair: [Doctor, Shift] | null = null;

    for (const shift of unfilled) {
      for (const doc of doctors) {
        if (shift.assigned_doctor_ids.includes(doc.doctor_id)) continue;
        const c = getCost(doc, shift);
        if (c < bestCost) {
          bestCost = c;
          bestPair = [doc, shift];
        }
      }
    }

    if (!bestPair || bestCost === Infinity) break;

    const [bestDoc, bestShift] = bestPair;
    bestShift.assigned_doctor_ids.push(bestDoc.doctor_id);
    assignments.get(bestDoc.doctor_id)!.push(bestShift);
  }

  return shifts;
}

/* ---------- Top-level schedule generation ---------- */

export function generateSchedule(
  doctors: Doctor[],
  year: number,
  month: number
): { shifts: Shift[]; holidays: string[] } {
  const { holidays, prefestivi } = getItalianHolidaysAndPrefestivi(year);
  const shifts = generateShiftsForMonth(year, month, holidays, prefestivi);
  assignShifts(doctors, shifts);
  return { shifts, holidays };
}
